lib/src/authentication/client.dart

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// ignore_for_file: import_of_legacy_library_into_null_safe
+
+import 'dart:io';
+
+import 'package:http/http.dart' as http;
+
+import '../http.dart';
+import '../log.dart' as log;
+import '../system_cache.dart';
+import 'token.dart';
+
+/// This client authenticates requests by injecting `Authentication` header to
+/// requests.
+///
+/// Requests to URLs not under [serverBaseUrl] will not be authenticated.
+class _AuthenticatedClient extends http.BaseClient {
```

Comment:
> I imagine the Future<http.StreamedResponse> send(http.BaseRequest request) method would add the token header if the host of the request has a token, and if we have a 401 we log the error message.

Pub server 'https://serverA.com/' might redirect to 'https://serverB.com/download/pkg.tar.gz'. In this case authentication header should not be attached to the request because of security reasons as described in [specs document](https://docs.google.com/document/d/1cJygfURx3hpX4oengdiDmpYuSWRQ3w3CnTAvGBnJSlY/edit?resourcekey=0-E30NU2oL_-F0nnIA6--Yzw#) "Client / Server authentication Protocol" section:

> - The requested URL is invoked for listing versions or fetching tarball from the server for which the credential is registered.
>   - Example: If a tokenA is registered for https://serverA.com and the client makes a request to list versions from serverB which says to download the tarballs to https://serverA.com/... then tokenA must not be attached to the request.

Reply:
Discussed this with @jonasfj and I think this is for another day.

Your solution is good for now!